### ISSUE TITLE
Betweenness with cutoff - wrong output

### ISSUE THREAD

**Issue by ibestvina** [igraph/igraph#1316]
igraph version 0.7.1

Code:

```
import igraph as ig
g = ig.Graph()
g.add_vertices(4)
g.add_edge(0, 1, weight = 1)
g.add_edge(0, 2, weight = 2)
g.add_edge(1, 3, weight = 1)
g.add_edge(2, 3, weight = 2)

print(g.betweenness(weights='weight', directed=False, cutoff=2))
```

Output:
`[0.25, 1.0, 0.0, 0.25]`

Expected output:
`[0.0, 1.0, 0.0, 0.0]`
because only the length of the shortest path between 0 and 2 (going through node 1) is below the cutoff.

**Comment by szhorvat** [comment]
> To be honest, I'm a bit skeptical about the whole betweenness estimation idea based on total path length cutoffs.

I just wanted to note that estimation is not the only application for what this function does. Range-limited betweenness is a very useful concept on its own that I have used many times in spatial networks. igraph provides this feature uniquely—it should be preserved.

As for estimation, I should finally take the time to read these:

 - https://journals.aps.org/prl/abstract/10.1103/PhysRevLett.105.038701
 - https://arxiv.org/abs/1111.5382

**Comment by ntamas** [comment]
Even if that is the case, we should check whether it's implemented correctly in igraph. I was looking at the source code a bit this morning, and it seems like we have basically implemented the Brandes algorithm for betweenness and then we simply stop the BFS when we reach the cutoff. It seems okay to me, but YMMV.

**Comment by szhorvat** [comment]
> we should check whether it's implemented correctly in igraph

Yes, absolutely. The result from the OP does not look correct.

Here's a simpler example. This is a weighted graph where all the weights are 1.0:

![image](https://user-images.githubusercontent.com/1212871/73725564-f3f15000-472d-11ea-9843-faa10cd40a8d.png)

The betweenness values are clearly 0, 3, 4, 3, 0.

```
In[175]:= IGBetweennessEstimate[g2, Infinity (* cutoff *)]
Out[175]= {0., 3., 4., 3., 0.}
```

There are four shortest paths going through the middle vertex. The longest one of these has length 4. The is just one path of length 4.

Therefore, I would expect the betweenness of the middle vertex to drop to 3 once the cutoff crosses 4.0. However, it actually drops to 3 only when the cutoff crosses 3.0, not 4.0.

```
In[178]:= IGBetweennessEstimate[g2, 3.01]
Out[178]= {0., 3., 4., 3., 0.}

In[179]:= IGBetweennessEstimate[g2, 2.99]
Out[179]= {0., 2., 3., 2., 0.}
```

**Comment by vtraag** [comment]
I was already looking at the betweenness calculations earlier (PR #1209, #1199) while @iosonofabio was looking the cutoff specifically (PR #1187). Perhaps just assign one of us to the problem?

**Comment by ntamas** [comment]
Assigned both of you -- toss a coin :)

**Comment by szhorvat** [comment]
Just a gentle reminder, as I find this bug quite concerning.

**Comment by iosonofabio** [comment]
I'm looking at it, hope to make some progress. Since I have another paper I'm doing revisions for, it might take a few days

**Comment by iosonofabio** [comment]
Ok let's start from the beginning because I'm not very smart, please bear with me.

First of all, our C function `igraph_betweenness` just calls `igraph_betweenness_estimate` with a negative cutoff. So looks like Python only exposes the latter but with a default negative cutoff, replicating the thin C behaviour - btw that seems inconsistent to Mathematica which calls it `IGBetweennessEstimate` instead of the expected `IGBetweenness`, but that's for another day.


Second, the OP uses weights but we have two separate functions in C, so let's start without weights first:
```python
import igraph as ig
g = ig.Graph()
g.add_vertices(4)
g.add_edge(0, 1)
g.add_edge(0, 2)
g.add_edge(1, 3)
g.add_edge(2, 3)

print(g.betweenness(directed=False))
[0.5, 0.5, 0.5, 0.5]

print(g.betweenness(directed=False, cutoff=3))
[0.5, 0.5, 0.5, 0.5]

print(g.betweenness(directed=False, cutoff=2))
[0.5, 0.5, 0.5, 0.5]

print(g.betweenness(directed=False, cutoff=1.01))
[0.5, 0.5, 0.5, 0.5]

print(g.betweenness(directed=False, cutoff=1))
[0.0, 0.0, 0.0, 0.0]   <---
```
For starters, this does not seem like the correct behaviour to me: the docs say that the cutoff is **The maximal length of paths that will be considered**. With a cutoff of 1.01, why are paths of length 2 still considered?

The key line in the unweighted function is inside the BFS:
```C
            if (cutoff > 0 && distance[actnode] >= cutoff + 1) {
                continue;
            }
```
So with a cutoff of 1.01, distances of 2 are still taken. I understand that without weights one shouldn't use floating cutoffs, but still - our docs are either wrong or confusing and igraph does not throw an error: we should fix that by either:
 1. changing the BFS line to the more appropriate:
```C
            if (cutoff > 0 && distance[actnode] > cutoff) {
                continue;
            }
```
which would be consistent with the docs but perhaps prone to rounding errors, or

 2. changing the BFS to:
```C
            if (cutoff > 0 && distance[actnode] > 0.00000000001 + cutoff) {
                continue;
            }
```
which would be less elegant and slightly less consistent but less prone to rounding errors, or

 3. throw an error when non-int cutoff are used with the unweighted function, as that's unexpected user behaviour. We should then also document it across all languages.

Third, the OP uses weights so let's get back to that but let's use weights of 1 for everyone for now:
```python

import igraph as ig
g = ig.Graph()
g.add_vertices(4)
g.add_edge(0, 1, weight=1)
g.add_edge(0, 2, weight=1)
g.add_edge(1, 3, weight=1)
g.add_edge(2, 3, weight=1)

print(g.betweenness(weights='weight', directed=False))
[0.5, 0.5, 0.5, 0.5]

g.betweenness(weights='weight', directed=False, cutoff=1.01)
[0.5, 0.5, 0.5, 0.5]

g.betweenness(weights='weight', directed=False, cutoff=1.0)
[0.0, 0.0, 0.0, 0.0]
```
That seems wrong but in the same way the unweighted function is wrong. Except that now with weights floating cutoffs should probably allowed (?) so the line with `cutoff=1.01` is a bona fide bug I think.

Fourth comes the question with the original, uneven weights:
```python

import igraph as ig
g = ig.Graph()
g.add_vertices(4)
g.add_edge(0, 1, weight=1)
g.add_edge(0, 2, weight=2)
g.add_edge(1, 3, weight=1)
g.add_edge(2, 3, weight=2)

print(g.betweenness(directed=False))
[0.5, 0.5, 0.5, 0.5]  <--- the unweighted version without cutoff obviously works

g.betweenness(weights='weight', directed=False)
[0.5, 1, 0.0, 0.5] <--- this is the result reported by the original user with cutoff=300: correct

g.betweenness(weights='weight', directed=False, cutoff=2.01)
[0.5, 1.0, 0.0, 0.5] <-- wrong answer (paths of length 3 are still kept), see bug of the unweighted function

g.betweenness(weights='weight', directed=False, cutoff=2.0)
[0.25, 1.0, 0.0, 0.25] <-- wrong answer, perhaps a different bug? spoiler: probably not.
```
So it seems like in addition to the bug in the unweighted function - which is ALSO present in the weighted function, see second to last example, there is another one that is specific to the weighted function, see last example. Or maybe it is the same but causes stranger things when weights are evoked.

Anyway let's look into that weighted function - which is internal to igraph, not part of the API. The variable `cutoff` is used in exactly one line in that function:
```C
            long int minnei = igraph_2wheap_max_index(&Q);
            ...
            if (cutoff > 0 && VECTOR(dist)[minnei] >= cutoff + 1.0) {
                continue;
            }
```
so even without knowing anything about the function - which is very sparsely commented, perhaps because it was never designed to be part of the API - it seems reasonable to assume the two bugs are both related to this line of code. Indeed, with `cutoff=2.0` paths of (weighted) length 3 are still considered according to this code snippet, but they shouldn't.

**tl;dr: shall we change this line with the cutoff in both the unweighted and weighted functions to take out the + 1.0 and/or include a small offset to deal with rounding errors?**

**Comment by vtraag** [comment]
@iosonofabio thanks for the extensive report. I think it relates to #1199 and #1209. I will try to take a closer look, but it's difficult to find time to sit down.

**Comment by ntamas** [comment]
Regarding the `+ 1.0` part in the code -- it might be a remnant of a trick that we used to use in several places in the codebase.

In many algorithms, you need a vector that maps from vertex IDs to some distance related to that vertex (this is typically the case in BFS, Dijkstra's algorithm etc), and the default value of the vector should be a value that represents "vertex not visited yet". When you have (non-negative) distances, a natural choice for such a default value would be -1. However when you use `calloc()` et al to allocate memory, the vector is filled with zeros by default, so either we spend O(n) time to fill the vector with -1, or we simply say that we store the distance _plus one_ in the vector, and we keep zero as the marker for the case when the vertex was not visited yet.

This worked fine for integer weights (apart from cases when we forgot to add or subtract 1 in the appropriate places and then we spent hours looking for the cause of a bug). However, if the weights are floats close to zero, adding 1 to them degrades the accuracy of the result. For instance, 1e-11 can be represented with an absolute error of -3.99e20 in floats, but 1 + 1e-11 is rounded down to 1. ([You can test it here](https://www.h-schmidt.net/FloatConverter/IEEE754.html)). So, we eventually stopped doing this and we have already re-written some parts of the codebase where accuracy suffered because of this premature "optimization". It could be the case that the "+1" you have found is somehow related to a similar hack. Either way, if you see tricks like this anywhere in the code, please file an issue about it so we can clean it up.

**Comment by vtraag** [comment]
Indeed some of the numerical issues in #1199 arose exactly because of this issue. As far as I'm concerned we can rewrite it to use -1 as a starting point, that will surely make the code a lot more easily readable. I'm not sure whether it will immediately solve the issue at hand, but it will be an improvement nonetheless.

**Comment by iosonofabio** [comment]
Ok then if you guys are on board I'll kill the +1 trick here and raise a flag if I see it elsewhere.

Regarding the second problem, i.e. adding to stack before checking for cutoff in the weighted function, I guess in absence of a better lead I'll test with/without swapping the two lines and see what happens. Will keep you posted.

**Comment by iosonofabio** [comment]
Btw, the same trick is used in a bunch of places across the file, e.g. closeness estmates. Shall I refactor all of those as well?

**Comment by szhorvat** [comment]
> Btw, the same trick is used in a bunch of places across the file, e.g. closeness estmates. Shall I refactor all of those as well?

If you do it, I'd suggest making it a separate PR.

**Comment by iosonofabio** [comment]
That's a no. Any other opinions on the table? If not I will only change those two functions and the rest will just be inconsistent, which is fine for me.

**Comment by vtraag** [comment]
I think it's fine to refactor all those functions where that appears and make a PR for that. But I think what @szhorvat means is that fixing this issue about the cutoff should be addressed in a separate PR. That also makes sense to me.

**Comment by szhorvat** [comment]
Sorry for being unclear.

If you have time to also address this for the closeness, that would be great!

I simply wanted to suggest to have separate PRs for betweenness-related and for closeness-related changes, if not too inconvenient (i.e. if the two PRs would not conflict—I do not think they will). If something breaks, it will be easier to keep track of things that way.

**Comment by szhorvat** [comment]
Here's an exact Mathematica implementation for *unweighted* betweenness with cutoff. I used this to verify PR #1392. The notebook (containing the tests) is also attached.

```mathematica
findAllShortestPaths[g_, s_, t_] :=
 With[{d = GraphDistance[g, s, t]},
  FindPath[g, s, t, {d}, All]
  ]

betweenness[g_, cutoff_ : Infinity] :=
 Module[{paths, counts, res},
  counts = Catenate@Catenate@Table[
      paths = findAllShortestPaths[g, s, t];
      paths = Select[paths, Length[#] <= cutoff + 1 &];
      {#, 1/Length[paths]} & /@ Join @@ paths[[All, 2 ;; -2]],
      {s, VertexList[g]}, {t, VertexList[g]}
      ];
  res = Lookup[
    Total /@ GroupBy[counts, First -> Last],
    VertexList[g],
    0
    ];
  If[UndirectedGraphQ[g], res/2, res]
  ]
```

[Betweenness with cutoff.nb.zip](https://github.com/igraph/igraph/files/4705493/Betweenness.with.cutoff.nb.zip)
